Various smaller lua_api documentation updates

Wuzzy2 opened the pull request:
This PR fixes several smaller documentation issues at once because posting one PR for every tiny documentation fix is a nightmare.

Fixes #7897.
Fixes #6492.
Fixes #11956.
Fixes #12113.
Fixes #12107.
Fixes #12361.

fluxionary commented:
i've got a small list of very minor issues w/ the lua api doc, if you wouldn't mind adding the fixes to this PR:

* active object reference
  * https://github.com/minetest/minetest/blob/cf19167e99d2212e6e8727ce1a277d1bb7cd719c/doc/lua_api.txt#L7222 `get_formspec_prepend(formspec)` -- doesn't actually take an argument
  * https://github.com/minetest/minetest/blob/cf19167e99d2212e6e8727ce1a277d1bb7cd719c/doc/lua_api.txt#L7282 `hud_get_hotbar_itemcount` missing parens
  * https://github.com/minetest/minetest/blob/cf19167e99d2212e6e8727ce1a277d1bb7cd719c/doc/lua_api.txt#L7285 `hud_get_hotbar_image` missing parens
  * https://github.com/minetest/minetest/blob/cf19167e99d2212e6e8727ce1a277d1bb7cd719c/doc/lua_api.txt#L7288 `hud_get_hotbar_selected_image` missing parens
* environment access
  * https://github.com/minetest/minetest/blob/cf19167e99d2212e6e8727ce1a277d1bb7cd719c/doc/lua_api.txt#L5489 `minetest.get_node_light(pos, timeofday)` should be `minetest.get_node_light(pos[, timeofday])`
* minetest.get_player_information(player_name)
  * https://github.com/minetest/minetest/blob/cf19167e99d2212e6e8727ce1a277d1bb7cd719c/doc/lua_api.txt#L4974 missing `,` after `lang_code = "fr"`

Wuzzy2 commented:
@fluxionary: Done.

Desour commented on doc/lua_api.txt:
```diff
@@ -5428,6 +5439,7 @@ Authentication
     * The player needs to be online for this to be successful.
 
 * `minetest.get_auth_handler()`: Return the currently active auth handler
+    * Must be called *after* load time, returns `nil` otherwise
```
Wouldn't it otherwise return `core.builtin_auth_handler`?

(Still, it's probably good to tell to only call this after the auth handler is properly set up (as long as the modder doesn't know what they're doing). :+1:)

Desour replied:
See https://github.com/minetest/minetest/blob/b7359f5fa93f531986c86b1ced20bb4584593b0a/builtin/game/auth.lua#L163-L165 btw..

Something like this:
```suggestion
    * Must be called *after* load time, to ensure that any custom auth handler was
      already registered.
```

Desour commented on doc/lua_api.txt:
```diff
@@ -6911,16 +6924,65 @@ of the `${k}` syntax in formspecs is not deprecated.
-    * returns `true` on success
+* `to_table()`:
+    * Returns metadata as a metadata table (see below), if possible
+    * Return `nil` on failure
```
```suggestion
    * Returns `nil` on failure
```

Desour commented on doc/lua_api.txt:
```diff
@@ -6911,16 +6924,65 @@ of the `${k}` syntax in formspecs is not deprecated.
+    * Imports metadata from a metadata table
+    * If `data` is a metadata table (see below),
+      the metadata it represents will replace all metadata
+      of this MetaDataRef object
```
Is there a reason why you're using so short lines?

Desour commented on doc/lua_api.txt:
```diff
@@ -6911,16 +6924,65 @@ of the `${k}` syntax in formspecs is not deprecated.
+    * inventory table keys are inventory list names
+    * inventory table values are item tables
+    * item table keys are slot IDs (starting with 1)
+    * item table values are itemstrings
+    * empty item slots *must* be `""`
```
Anything convertible to an ItemStack is fine for `from_table` (i.e. also an ItemStack, a table or `nil`). (Even `nil` should be fine, as `read_items` iterates with `lua_next`. Maybe discourage it though.)

For `to_table`, it should be ItemStacks, as `push_items` calls `LuaItemStack::create`. I haven't tested though.

Desour replied:
Also, the description could be simpler here, i.e.:
```
* `inventory` is a table that maps listnames to itemlists
* An itemlists is a list of `ItemStack`s, or anything convertible to it. The index (beginning at 1) specifies the inventory list slot.
```

Desour commented on doc/lua_api.txt:
```diff
@@ -6911,16 +6924,65 @@ of the `${k}` syntax in formspecs is not deprecated.
+    metadata_table = {
+        -- metadata fields (key/value store)
+        fields = {
+            infotest "Container",
```
```suggestion
            infotest = "Container",
```

Desour commented on doc/lua_api.txt:
```diff
@@ -8569,15 +8633,17 @@ Used by `minetest.register_node`.
-        -- default: nil
+        -- * `pos`: node position
+        -- * `oldnode`: node table of node before it was deleted
+        -- * `oldmeta` metadata of node before it was deleted, as a metadata table
+        -- * `drops` is a table of `ItemStack`s, so any metadata to be preserved can
```
```suggestion
        -- * `oldmeta`: metadata of node before it was deleted, as a metadata table
        -- * `drops`: a table of `ItemStack`s, so any metadata to be preserved can
```

Desour reviewed (commented):
It doesn't look like #11292 and #11736 were addressed.

Wuzzy2 commented:
Updated! I accidentally listed the two issues, I removed them from the description now. `get_pos` is handled in #13239 instead.

I do not agree with all change requests. The "simpler" description of metadata tables is too vague IMHO.

sfan5 reviewed (approved):
lgtm

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -6911,16 +6925,63 @@ of the `${k}` syntax in formspecs is not deprecated.
 
+### Metadata tables
+
+A MetaDataRef can be converted to and from a metadata table,
+which is metadata represented in a table form. It is used
+to serialize and deserialize metadata.
+The methods `from_table` and `to_table` use metadata tables, as
+well as a few callback functions.
```
What do you mean by "a few callback functions"? I do not understand what you want to say with the second part of this sentence.

SmallJoker replied:
Wouldn't this short introduction suffice?

```suggestion
Metadata tables represent MetaDataRef in a Lua table form (see `from_table`/`to_table`).
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -6911,16 +6925,63 @@ of the `${k}` syntax in formspecs is not deprecated.
-    * See [Node Metadata] for an example
-    * returns `true` on success
+* `to_table()`:
+    * Returns metadata as a metadata table (see below), if possible
+    * Returns `nil` on failure
```
What does "if possible" refer to? When it it not possible? Proposed alternative:

```suggestion
    * Returns a Metadata table (see below) or `nil` on failure.
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -6911,16 +6925,63 @@ of the `${k}` syntax in formspecs is not deprecated.
+    * all values are stored as strings
+    * numbers must be converted to strings first
+* `inventory` (for NodeMetaRef only): A node inventory in table form
+    * inventory table keys are inventory list names
+    * inventory table values are item tables
+    * item table keys are slot IDs (starting with 1)
+    * item table values are ItemStacks or anything convertible to it
```
"ItemStacks or anything convertible to it" is not reliable information. They *are* ItemStacks. Suggestion:

```suggestion
    * Key: string, name of the inventory list
    * Value: list of `ItemStack`, same as `InvRef:get_list`
```

Wuzzy2 commented:
Updated, but there are still open questions:

> "ItemStacks or anything convertible to it" is not reliable information. They are ItemStacks.

Hmmm, so technically, if you do `from_table`, itemstrings like `"default:stone 5"` are also accepted, which I discovered by experimenting. Is this OK to document this fact? Or is the itemstring support just an accidental side effect the Lua developer should not rely on? If the latter is true, then the metadata table examples are also wrong and need updating (as they all use strings).

The "few callback" functions I mentioned are the node functions `after_dig_node` and `preserve_metadata`. I removed this confusing wording now.

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -460,9 +460,10 @@ texture is overlaid on top of `cobble.png`.
+is required for `^`, `:` and `\`.
 
 Example: `cobble.png^[lowpart:50:color.png\^[mask\:trans.png`
+Or as Lua string: `"cobble.png^[lowpart:50:color.png\\^[mask\\:trans.png"`
```
```suggestion
Or as a Lua string: `"cobble.png^[lowpart:50:color.png\\^[mask\\:trans.png"`
```

rubenwardy commented:
Looks good apart from that + clarifications related to to/from_table ItemStack

SmallJoker commented:
> Hmmm, so technically, if you do `from_table`, itemstrings like `"default:stone 5"` are also accepted,

This is an implicit conversion when reading back the table. Minetest by itself always outputs `ItemStack` userdata but accepts itemstring too for `from_table` calls.. Perhaps you could add a note for this conversion into the `from_table` function description?

Desour commented:
>Perhaps you could add a note for this conversion into the `from_table` function description?

Correct me if I'm wrong, but all our API functions that require an ItemStack also accept anything convertible to an ItemStack, right?. So, maybe write that down in the ItemStack section.

Wuzzy2 commented:
Updated it again, I'm not sure if I got it right this time. \*shrug\*

SmallJoker reviewed (approved):
Forgot to approve. Looks fine.

Zughy commented:
@Wuzzy2 rebase needed

Wuzzy2 commented:
Rebase completed.

Wuzzy2 commented:
@Desour: Done.
